File: controlEatFoodWeb/backend/src/main/java/com/eatfood/control/util/CedulaValidator.java
package com.eatfood.control.util;

/**
 * Validación de la cédula de identidad ecuatoriana (10 dígitos):
 * <ul>
 *   <li>Provincia (2 primeros dígitos): 01–24, o 30 (ecuatorianos registrados en el exterior).</li>
 *   <li>Tercer dígito: 0–5 (personas naturales).</li>
 *   <li>Décimo dígito: verificador por módulo 10 con coeficientes 2,1,2,1,2,1,2,1,2
 *       (si el producto es mayor a 9 se le resta 9).</li>
 * </ul>
 */
public final class CedulaValidator {

    private CedulaValidator() {}

    /** ¿El valor tiene exactamente 10 dígitos numéricos? (forma de cédula, aún sin verificar). */
    public static boolean looksLikeCedula(String value) {
        return value != null && value.matches("\\d{10}");
    }

    /** Valida una cédula ecuatoriana completa (formato + provincia + dígito verificador). */
    public static boolean isValid(String value) {
        if (!looksLikeCedula(value)) return false;

        int province = Integer.parseInt(value.substring(0, 2));
        if ((province < 1 || province > 24) && province != 30) return false;

        int thirdDigit = value.charAt(2) - '0';
        if (thirdDigit > 5) return false;

        int sum = 0;
        for (int i = 0; i < 9; i++) {
            int digit = value.charAt(i) - '0';
            int product = (i % 2 == 0) ? digit * 2 : digit;
            if (product > 9) product -= 9;
            sum += product;
        }
        int expected = (10 - (sum % 10)) % 10;
        return expected == (value.charAt(9) - '0');
    }
}

File: controlEatFoodMovil/app/src/main/java/com/eatfood/control/mobile/util/CedulaValidator.kt
package com.eatfood.control.mobile.util

/**
 * Validación de la cédula de identidad ecuatoriana (10 dígitos).
 * Mismo algoritmo que el backend (CedulaValidator.java) y la web (utils/cedula.js):
 * provincia 01-24 o 30, tercer dígito 0-5 y dígito verificador por módulo 10.
 */
object CedulaValidator {

    /** ¿Tiene forma de cédula (exactamente 10 dígitos)? */
    fun looksLikeCedula(value: String?): Boolean =
        value != null && Regex("\\d{10}").matches(value)

    /** Valida formato, provincia y dígito verificador. */
    fun isValid(value: String?): Boolean {
        if (!looksLikeCedula(value)) return false
        val v = value!!

        val province = v.substring(0, 2).toInt()
        if (province !in 1..24 && province != 30) return false

        if (v[2] - '0' > 5) return false

        var sum = 0
        for (i in 0 until 9) {
            var product = (v[i] - '0') * if (i % 2 == 0) 2 else 1
            if (product > 9) product -= 9
            sum += product
        }
        val expected = (10 - (sum % 10)) % 10
        return expected == v[9] - '0'
    }
}
